RFC-0009/RFC-0009-binary-exchange.md

```diff
@@ -0,0 +1,265 @@
+
+The implementation covers the protocol design, the BinX server implementation and the BinX exchange
+source implementation.
+
+### Binary Exchange Protocol
+
+The protocol consists of requests and replies that are implemented by the `BinRequest` and the `BinReply` classes defined
+in `BinRequestReply.h`. All three request types for data, acknowledge and delete share the same structure.
+A request consists of the following fields:
+|Field name| Type | Semantic |
+| ---- | ---- | ---- |
+|requestType| enum (DATA, ACK, DELETE) | The type of request |
+|getDataSize| bool | When true, requests the sizes of the remaining available data pages; when false requests data |
+|maxSizeOctets| uint64_t | The maximum allowed size of the data pages|
+|maxWaitMicroSeconds|uint64_t| The maximum wait time in microseconds|
+|taskId| std::string | The unique ID of the remote task from which data is requested|
+|bufferId| uint64_t | The ID of the buffer within the remote task|
+|pageToken| uint64_t | The sequence number of the data page that is requested|
+
+Acknowledge and delete requests don't need all fields and set `getDataSize`, `maxSizeOctets`,
+and `maxWaitMicroSeconds` to false and 0, respectively.
+
+A reply has the following fields:
+
+|Field name| Type | Semantic |
+| ---- | ---- | ---- |
+|replyType| enum (DATA, ACK, DELETE) | The type of reply |
+|status| enum (OK, NODATA, SERVER_ERROR, TIMEOUT) | The reply status |
+|bufferComplete| bool | True if the entire buffer has been transferred |
+|pageToken|uint64_t| The token (sequence number) of the first data page in the reply|
+|pageNextToken|uint64_t| The token (sequence number) of the next available page that can be requested|
+|remainingBytes| std::vector<uint64_t> | Contains the sizes of available pages when requested|
+|data| IOBuf | Contains the data payload |
+
+`BinRequest` and `BinReply` are in-memory representations. The serialization and deserialization on top
+of TCP is implemented by the `ServerSerializeHandler` and `ClientSerializeHandler`. On the server side,
+requests are read and replies are written and vice-versa on the client side. During serialization, a
+length field is prepended that is needed to correctly de-serialize the protocol data units.
+
+The implementation uses Meta's [Wangle](https://github.com/facebook/wangle) framework. 
+
+### Binary Exchange Server
+
+The binary exchange server is a self-contained component that is started when `binx.enabled` is configured.
+It implements the exchange service using the BinX protocol. Incoming requests are forwarded to the
+`TaskManager` and the result is marshalled into a reply and sent back.
+
+The binary exchange server is started in `PrestoServer::run()`. It uses a dedicated
+IO thread pool in order to not interfere with the HTTP IO thread pool. The CPU thread pool is shared
+with the HTTP exchange.
+
+#### Implementation Notes
+
+The BinX server uses Wangle. It consists of the following components that are implemented in
```

Comment:
Can you clarify how the Wangle stack is different from the Proxygen stack currently used by the HTTP implementation?
From a high-level, the asynchronous TCP socket handler `folly::AsyncServerSocket` is coming from folly in both stacks?

Reply:
Proxygen also uses Wangle underneath; at least portions of it. Wangle in turn uses folly's AsyncServerSocket. You are right, both stacks end up using folly's I/O mechanisms.
I'll add some clarifications.